library/psa_crypto.c

```diff
@@ -1189,36 +1189,48 @@ static psa_status_t psa_restrict_key_policy(
 
 /** Retrieve a slot which must contain a key. The key must have allow all the
  * usage flags set in \p usage. If \p alg is nonzero, the key must allow
- * operations with this algorithm. */
-static psa_status_t psa_get_key_from_slot( psa_key_handle_t handle,
+ * operations with this algorithm.
+ *
+ * On success, the access counter of the returned key slot is incremented by
+ * one. It is the responsibility of the caller to call
+ * psa_decrement_key_slot_access_count() when it does not access the key slot
+ * anymore.
+ */
+static psa_status_t psa_get_key_from_slot( mbedtls_svc_key_id_t key,
```

Comment:
For the names of the functions, I like the `acquire` verb. Otherwise:

. to replace psa_search_key_in_slots, not completely convinced by psa_acquire_key_in_memory. Other possibilities: psa_acquire_key_in_slot(s) or psa_acquire_loaded_key() ?
. to replace psa_get_key_slot maybe that psa_load_and_acquire_key() would be better than just psa_acquire_key() emphasizing the possible load. The weakness is that there may be no loading.
. an alternative to  psa_acquire_key_with_policy would then be psa_load_and_acquire_key_with_policy()

What do you think about those suggestions?

Reply:
* “Acquire key in slots” could mean a key that's acquired from slots or that's placed into slots. “Acquire loaded key” isn't ambiguous, but it looks like it excludes volatile keys. Maybe `psa_acquire_key_from_slots`? But I still prefer “memory”, because storage could have slots as well.
* Indeed “load and acquire” is a little misleading since it won't load already-loaded persistent keys and volatile keys. I do like having some disambiguation on just ”acquire“, but I don't think “load and acquire” is right. `psa_acquire_key_with_load`?

Reply:
ok thus last but not least. Eventually, I am not sure about `acquire` because when you acquire something you own it but here it is not the case, you don't own the key slot, somebody else can get a read access to it.

First regarding psa_get_key_slot(), from the point of view of the user, the user has got a key identifier and she wants the description of the key corresponding to that identifier. As key descriptions are stored in what we call key slots in the implementation, psa_get_key_slot()  seems at least a good base for the name.

Then, still from the point of view of the user whether the function has to load data for persistent storage or not does not really matter thus better not to change/tweak the base name for that.

On the contrary, the function also increments the key slot access counter and this is important for the user to be aware of that as it is her responsibility to decrement the counter when she's done with the key description. The access counter is there to prevent the key slot to be wiped out and being used for another purpose while a part of the library is making read accesses to it. Thus as you mentioned in [this comment](https://github.com/ARMmbed/mbedtls/pull/3547#discussion_r523406314) from the user point of view it is about locking the key slot to prevent it from being wiped out/reused while using it. Thus what about psa_get_and_lock_key_slot() ?

Then for psa_search_key_in_slots(), it would be psa_get_and_lock_key_slot_in_memory() and for psa_get_get_from_slot, psa_get_and_lock_key_slot_with_policy().

Basically I am proposing to replace `acquire` by `get_and_lock` in the function names you proposed. Otherwise, we would have  `psa_decrement_key_slot_access_count` → `psa_lock_key_slot`, `psa_increment_key_slot_access_count` → `psa_unlock_key_slot`.

What do you think?

Reply:
You can acquire a semaphore multiple times, so I don't think “acquire” is exclusive. But I'm ok with lock/unlock as well.

Reply:
Of the options discussed, I think I prefer the "aquire" / "release" verbs. I don't think that "load and aquire" is much more misleading than "aquire with load". I'd go with "load and aquire"  due to the interaction with the "with policy" name
As for the "from slots" / "from memory" question I have no strong feelings one way or the other. Maybe go with slots, since it's already used in the key slot functions?

So I'd probably go with:
psa_acquire_key_in_slots (/ psa_acquire_key_in_memory)
psa_load_and_acquire_key
psa_load_and_acquire_key_with_policy
psa_acquire_key_slot
psa_release_key_slot